src/search/merge_and_shrink/merge_and_shrink_algorithm.cc

```diff
@@ -55,8 +56,65 @@ MergeAndShrinkAlgorithm::MergeAndShrinkAlgorithm(
       log(utils::get_log_for_verbosity(verbosity)),
       main_loop_max_time(main_loop_max_time),
       starting_peak_memory(0) {
+    handle_shrink_limit_defaults();
+}
+
+void MergeAndShrinkAlgorithm::handle_shrink_limit_defaults() {
+    // If none of the two state limits has been set: set default limit.
+    if (max_states == -1 && max_states_before_merge == -1) {
+        max_states = 50000;
+    }
+
+    // If one of the max_states options has not been set, set the other
+    // so that it imposes no further limits.
+    if (max_states_before_merge == -1) {
+        max_states_before_merge = max_states;
+    } else if (max_states == -1) {
+        if (utils::is_product_within_limit(
+                max_states_before_merge, max_states_before_merge, INF)) {
+            max_states = max_states_before_merge * max_states_before_merge;
+        } else {
+            max_states = INF;
+        }
+    }
+
+    if (max_states_before_merge > max_states) {
+        max_states_before_merge = max_states;
+        if (log.is_warning()) {
+            log << "WARNING: "
+                << "max_states_before_merge exceeds max_states, "
+                << "correcting max_states_before_merge." << endl;
+        }
+    }
+
+    utils::verify_comparison(max_states, 1, greater_equal<>(),
+                             "Transition system size must be at least 1.");
+
+    utils::verify_comparison(max_states_before_merge, 1, greater_equal<>(),
+                             "Transition system size before merge must be at least 1.");
+
+    if (shrink_threshold_before_merge == -1) {
+        shrink_threshold_before_merge = max_states;
+    }
+
+    utils::verify_comparison(shrink_threshold_before_merge, 1, greater_equal<>(),
+                             "Threshold must be at least 1.");
+
+
+    if (shrink_threshold_before_merge > max_states) {
+        shrink_threshold_before_merge = max_states;
+        if (log.is_warning()) {
+            log << "WARNING: "
+                << "threshold exceeds max_states, "
+                << "correcting threshold." << endl;
+        }
+    }
+
+    // TODO are these asserts needed?
     assert(max_states_before_merge > 0);
     assert(max_states >= max_states_before_merge);
+    // TODO why is this assert comparing threshold to max_states_before_merge
+    // while the default handling checks the comparison of threshold to max_states
     assert(shrink_threshold_before_merge <= max_states_before_merge);
```

Comment:
It looks like this could be violated.

Reply:
Yes, with this call you can trigger the violation: `"astar(merge_and_shrink(shrink_strategy=shrink_bisimulation(greedy=false),merge_strategy=merge_sccs(order_of_sccs=topological,merge_selector=score_based_filtering(scoring_functions=[goal_relevance(),dfp(),total_order()])),label_reduction=exact(before_shrinking=true,before_merging=false),max_states=3k,threshold_before_merge=2k,max_states_before_merge=1k))"`

It is unclear to me why the assertion is there...